src/trace/trace.c

```diff
@@ -96,19 +116,144 @@ static inline void mtrace_event(const char *data, uint32_t length)
+		     entry_to_emit->suppress_count, entry_to_emit->entry_id);
+
+	memset(entry_to_emit, 0, sizeof(*entry_to_emit));
+
+	platform_shared_commit(trace, sizeof(*trace));
+}
+
+static void emit_recent_entries(uint64_t current_ts)
+{
+	struct trace *trace = trace_get();
+	struct recent_log_entry *entry_to_emit =
+		&trace->trace_core_context[cpu_get_id()].entry_to_emit;
+	struct recent_log_entry *recent_entries =
+		trace->trace_core_context[cpu_get_id()].recent_entries;
+	int i;
+
+	if (entry_to_emit->entry_id)
+		emit_recent_entry();
+
+	for (i = 0; i < CONFIG_TRACE_RECENT_ENTRIES_COUNT; i++) {
+		if (current_ts - recent_entries[i].message_ts >
+			CONFIG_TRACE_RECENT_TIME_THRESHOLD) {
+			if (recent_entries[i].suppress_count) {
+				prepare_to_emit(&recent_entries[i]);
+				emit_recent_entry();
+			} else {
+				memset(&recent_entries[i], 0, sizeof(recent_entries[i]));
+			}
+		}
+	}
+
+	platform_shared_commit(trace, sizeof(*trace));
+}
+
+/**
+ * \brief Runtime trace flood suppression
+ * \param log_level log verbosity level
+ * \param entry ID of log entry
+ * \param message_ts message timestamp
+ * \return false when trace is filtered out, true otherwise
+ */
+static bool trace_filter_flood(uint32_t log_level, uint32_t entry, uint64_t message_ts)
+{
+	struct trace *trace = trace_get();
+	struct recent_log_entry *recent_entries =
+		trace->trace_core_context[cpu_get_id()].recent_entries;
+	struct recent_log_entry *oldest_entry = recent_entries;
+	int i;
+
+	/* don't attempt to suppress debug messages using this method, it would be uneffective */
+	if (log_level >= LOG_LEVEL_DEBUG)
+		return true;
+
+	/* check if same log entry was sent recently */
+	for (i = 0; i < CONFIG_TRACE_RECENT_ENTRIES_COUNT; i++) {
+		/* in the meantime identify the oldest or an empty entry, for reuse later */
+		if (oldest_entry->first_suppression_ts > recent_entries[i].first_suppression_ts)
+			oldest_entry = &recent_entries[i];
+
+		if (recent_entries[i].entry_id == entry) {
+			if (message_ts - recent_entries[i].first_suppression_ts >=
+			    CONFIG_TRACE_RECENT_MAX_TIME) {
+				/* entry was suppressed long enough, time to release it */
+				if (recent_entries[i].suppress_count)
+					prepare_to_emit(&recent_entries[i]);
+				else
+					memset(&recent_entries[i], 0, sizeof(recent_entries[i]));
+
+				platform_shared_commit(trace, sizeof(*trace));
+				return true;
+			}
+			if (message_ts - recent_entries[i].message_ts <
+			    CONFIG_TRACE_RECENT_TIME_THRESHOLD) {
+				recent_entries[i].suppress_count++;
+				/* refresh suppression timer */
+				recent_entries[i].message_ts = message_ts;
+
+				platform_shared_commit(trace, sizeof(*trace));
+				return false;
+			}
+
+			if (recent_entries[i].suppress_count)
+				prepare_to_emit(&recent_entries[i]);
+			else
+				memset(&recent_entries[i], 0, sizeof(recent_entries[i]));
+			platform_shared_commit(trace, sizeof(*trace));
+			return true;
+		}
+	}
+
+	if (oldest_entry->entry_id && oldest_entry->suppress_count)
```

Comment:
an inline comment here is needed to so we understand why we emit oldest here.

Reply:
done